cli/src/main/java/com/box/l10n/mojito/cli/command/utils/SarifFileGenerator.java

```diff
@@ -0,0 +1,145 @@
+    SarifBuilder sarifBuilder = new SarifBuilder();
+    Map<String, AssetExtractorTextUnit> nameToAssetTextUnitMap =
+        assetExtractionDiffs.stream()
+            .flatMap(diff -> diff.getAddedTextunits().stream())
+            .collect(Collectors.toMap(AssetExtractorTextUnit::getName, x -> x));
+    for (CliCheckResult checkFailure : cliCheckerFailures) {
+      ResultLevel resultLevel = checkFailure.isHardFail() ? ResultLevel.ERROR : ResultLevel.WARNING;
+      sarifBuilder.addRun(checkFailure.getCheckName(), infoUri, this.gitInfo.getCommit().getId());
+      for (Map.Entry<String, CliCheckResult.CheckFailure> entry :
+          checkFailure.getNameToFailuresMap().entrySet()) {
+        String source = entry.getKey();
+        CliCheckResult.CheckFailure resultCheckFailure = entry.getValue();
+        AssetExtractorTextUnit assetExtractorTextUnit = nameToAssetTextUnitMap.get(source);
+        CheckerRuleId ruleId = resultCheckFailure.ruleId();
+        if (hasUsages(assetExtractorTextUnit)) {
+          sarifBuilder.addResultWithLocations(
+              ruleId.toString(),
+              resultLevel,
+              convertToTitleCase(checkFailure.getCheckName()),
+              resultCheckFailure.failureMessage(),
+              getUsageLocations(
+                  assetExtractorTextUnit, supportedFileExtensions, ruleId.isCommentRelated()));
+        } else {
+          sarifBuilder.addResultWithoutLocation(
+              ruleId.toString(),
+              resultLevel,
+              convertToTitleCase(checkFailure.getCheckName()),
+              resultCheckFailure.failureMessage());
+        }
+      }
+    }
+
+    return sarifBuilder.build();
+  }
+
+  static boolean hasUsages(AssetExtractorTextUnit assetExtractorTextUnit) {
+    return assetExtractorTextUnit != null
+        && assetExtractorTextUnit.getUsages() != null
+        && !assetExtractorTextUnit.getUsages().isEmpty();
+  }
+
+  static List<Location> getUsageLocations(
+      AssetExtractorTextUnit assetExtractorTextUnit,
+      String[] supportedFileExtensions,
+      boolean isCommentRelatedCheck) {
+    return assetExtractorTextUnit.getUsages().stream()
+        .map(
+            usage -> {
+              int colonIndex = usage.lastIndexOf(':');
+              if (colonIndex == -1) {
+                return null;
+              }
+
+              try {
+                String fileUri = usage.substring(0, colonIndex);
+                int startLineNumber = Integer.parseInt(usage.substring(colonIndex + 1));
+
+                if (!isCommentRelatedCheck) {
+                  return new Location(fileUri, startLineNumber);
+                }
+
+                int fullStopIndex = fileUri.lastIndexOf('.');
+                if (fullStopIndex == -1) {
+                  return new Location(fileUri, startLineNumber);
+                }
+
+                String fileExtension = fileUri.substring(fullStopIndex + 1);
```

Comment:
Guava Files.getFileExtension might be easier to follow here: https://guava.dev/releases/19.0/api/docs/com/google/common/io/Files.html#getFileExtension(java.lang.String)

Reply:
Done